### PR TITLE
Make PlanOptimizersStatsCollector thread-safe

### PR DESCRIPTION
**Pull request by Dith3r**
## Description

Fix https://github.com/trinodb/trino/issues/16960

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java [hunk 1]
```diff
@@ -18,15 +18,15 @@
...
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 public class PlanOptimizersStatsCollector
 {
-    private final Map<Class<?>, QueryPlanOptimizerStats> stats = new HashMap<>();
+    private final Map<Class<?>, QueryPlanOptimizerStats> stats = new ConcurrentHashMap<>();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:29]: The class was extracted from `IterativeOptimizer` in https://github.com/trinodb/trino/pull/16365
we perhaps can still be OK with additional overhead this brings

if we want to make this class thread-safe, then

- mark it as `@Threadsafe`
- `QueryPlanOptimizerStats` should become thread-safe too
- we no longer need `io.trino.sql.planner.iterative.IterativeOptimizer.Context#iterativeOptimizerStatsCollector` as a separate field, and we don't need `io.trino.execution.querystats.PlanOptimizersStatsCollector#add` method.

(I am not totally sure though whether this is the way to go.)

cc @martint

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:32]: > add is used when statistics from a single iterative optimizer are merged to collect statistics for a query.

this i know

however, now that `PlanOptimizersStatsCollector` is thread-safe, why create a new instance of that thing in `IterativeOptimizer.Context`?
... actually, it's not directly related to thread-safety -- it was redundant even before, so we can keep it as a follow-up

Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:32]: It was created to mirror the behavior that was implemented before. When the iterative optimizer exceeds the time limit, we log the top N rules that the current iterative optimizer has executed, as opposed to a top-level `PlanOptimizersStatsCollector` which collects statistics from rules across all optimizers.

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java:32]: That's a good point.

OTOH, when iterative optimizer fails (or query is cancelled), the query completion event top N rules won't include rules from the last iterative optimizer run, which are usually the most important ones.

can you create an issue & follow-up on this?

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/16972 , https://github.com/trinodb/trino/issues/15430

**Pull request merged**